rfcs/experimental/flow_graph_wait_try_put_and_wait/README.md

```diff
@@ -0,0 +1,604 @@
+
+The ``try_put_and_wait`` is expected to exit when all of the red (including partially red) tasks are completed. It may require some amount of blue tasks
+to be completed as well, e.g. to execute previously stored tasks in the ``function_node``'s queue. 
+
+Completion of all blue tasks as in ``wait_for_all`` is not guaranteed.
+
+## Feature Design
+
+This feature is implemented by creating an instance of special private class ``message_metainfo`` for each input message in ``try_put_and_wait``. This instance
+wraps the pointer to the dedicated ``wait_context`` object representing the reference counter for single input message.
+
+``message_metainfo`` is broadcasted through the graph together with the message itself. The actual value and the type of the message can be changed during the computation, but the stored 
+metainformation is preserved.
+
+The reference counter in the ``wait_context`` inside of the ``message_metainfo`` is increased in the following cases:
+* The task associated with the computations of the ``try_put_and_wait`` input message is created (*).
+* The item associated with the computations of the desired input is stored in the buffering node or in the internal buffer of the other node type.
+* When the ``continue_msg`` with the non-empty associated waiter is received by ``continue_node``. In this case, the node buffers the ``message_metainfo``s received from each
+  predecessor and increases the underlying reference counter to prolonge the wait until signals from each predecessor would be received. 
+
+The reference counter is decreased when:
+* The task associated with the computations of the ``try_put_and_wait`` input is finalized. If the output of the task should be propagated to the successors of the node, it is done
+  from the task body and hence the reference counter would be decreased only after creating the task for each successor and increasing the reference counter for them.
+* The item associated with the computations is taken from the buffering node or from the internal buffer.
+* When the desired number of signals from the predecessors was received by the ``continue_node``. This case is equivalent to retrieving the set of buffered ``message_metainfo``s received previously
+  from each predecessor. 
+
+``message_metainfo`` class may be reused in the future to support additional use-cases when it is required to push additional information about the input message through the graph. E.g. supporting
+priorities for single messages - the corresponding priority tag can be assigned as part of the metainfo.
+
+Metainformation class supports containing multiple reference counters at the same time to support joining messages with different associated reference counters in the ``join_node``. 
+See [separate section](#details-about-metainformation-class) for more details.
+
+Implementation-wise, processing the metainformation is exposed by adding new internal virtual member functions to various Flow Graph instance:
```

Comment:
I would add a note that the functions exposed below are details of implementation which might change over time and therefore should not be used directly.